### PR TITLE
feat: support query on resource IDs

### PR DESCRIPTION
**Pull request by smira**
Allow List/WatchKind results to be filtered on resource IDs.

See https://github.com/cosi-project/specification/pull/6

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

### CODE REVIEW

File: pkg/resource/id_query.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+// IDQuery is the query on the resource ID.
+type IDQuery struct {
+	Regexp *regexp.Regexp
+}
+
+// Matches if the resource ID matches the ID query.
+func (query IDQuery) Matches(md Metadata) bool {
+	if query.Regexp == nil {
+		return true
```
Comment (frezbo) [inline pkg/resource/id_query.go:19]: match true if it's nil? :thinking:

File: pkg/resource/metadata_test.go [hunk 2]
```diff
@@ -187,7 +187,7 @@ finalizers:
 	assert.True(t, md.Equal(in))
 }
 
-var ts, _ = time.Parse(time.RFC3339, "2021-06-23T19:22:29Z")
+var ts, _ = time.Parse(time.RFC3339, "2021-06-23T19:22:29Z") //nolint:errcheck
```
Comment (DmitriyMV) [inline pkg/resource/metadata_test.go:190]: Linter is probably right here - we do want to ensure that `err` is `nil` I suggest something like this
```suggestion
var ts = func() time.Time {
	res, err := time.Parse(time.RFC3339, "2021-06-23T19:22:29Z")
	if err != nil {
		panic(err)
	}

	return res
}()
```

**Review by frezbo — APPROVED** [review]

File: pkg/state/protobuf/server/server.go [hunk 3]
```diff
@@ -75,6 +75,17 @@ func (server *State) List(req *v1alpha1.ListRequest, srv v1alpha1.State_ListServ
 				opts = append(opts, state.WithLabelQuery(labelOpts...))
 			}
 		}
+
+		if req.Options.GetIdQuery() != nil {
+			var idOpts []resource.IDQueryOption
```
Comment (DmitriyMV) [inline pkg/state/protobuf/server/server.go:80]: Redundant declaration?

File: pkg/resource/id_query.go (hunk 1 above)
Comment (smira) [reply pkg/resource/id_query.go:19]: yeah, it means that no query was set

File: pkg/state/protobuf/server/server.go [hunk 4]
```diff
@@ -75,6 +75,17 @@ func (server *State) List(req *v1alpha1.ListRequest, srv v1alpha1.State_ListServ
...
+
+			idOpts, err := ConvertIDQuery(req.Options.GetIdQuery())
```
Comment (DmitriyMV) [inline pkg/state/protobuf/server/server.go:82]: This variant and the one below are different. Is there any reason for this?

File: pkg/resource/metadata_test.go (hunk 2 above)
Comment (DmitriyMV) [reply pkg/resource/metadata_test.go:190]: Okay, although I think if we can check things during assignments , we should do this. I usually just wrap the expression in `func must[T any](v T, err error) T`. Actually we should probably have `testutils` in `gen` so we could stop writing `must` for every package.

File: pkg/state/protobuf/server/server.go (hunk 4 above)
Comment (smira) [reply pkg/state/protobuf/server/server.go:82]: I guess I was refactoring it and missed it, not real difference, I will clean it up, thanks

**Review by Unix4ever — APPROVED** [review]

**Pull request merged**